projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php

```diff
@@ -247,6 +248,93 @@ private function maybe_gate_access_for_user_if_post_tier( $post_id, $user_abbrev
 		return $this->maybe_gate_access_for_user_if_tier( $tier_id, $user_abbreviated_subscriptions );
 	}
 
+	/**
+	 * Get all plans id that make access valid for a post with this tier id.
+	 *
+	 * @param int $tier_id Tier ID.
+	 *
+	 * @return array|WP_Error
+	 */
+	public static function get_valid_plan_ids_for_tier( $tier_id ) {
+		// Valid plans are:
+		// - monthly tiers with same currency and price same or higher than original tier
+		// - yearly tiers higher than related yearly plan or 12 times price of the original tier
+		$valid_plan_ids = array();
+
+		$all_plans = \Jetpack_Memberships::get_all_plans();
+
+		// Let's get the current tier
+		$tier = null;
+		foreach ( $all_plans as $post ) {
+			if ( $post->ID === $tier_id ) {
+				$tier = $post;
+				break;
+			}
+		}
+
+		if ( $tier === null ) {
+			// We have an error
+			return new WP_Error( 'related-plan-not-found', 'The plan related to the tier cannot be found' );
+		}
+
+		$tier_price      = self::find_metadata( $tier, 'jetpack_memberships_price' );
+		$tier_currency   = self::find_metadata( $tier, 'jetpack_memberships_currency' );
+		$tier_product_id = self::find_metadata( $tier, 'jetpack_memberships_product_id' );
+
+		if ( $tier_price === null || $tier_currency === null || $tier_product_id === null ) {
+			// There is an issue with the meta
+			return new WP_Error( 'wrong-data-plan-not-found', 'The plan related to the tier is missing data' );
+		}
+
+		$valid_plan_ids[] = $tier_id;
+
+		$tier_price = floatval( $tier_price );
+
+		// At this point we know the post is
+		$annual_tier = null;
+		foreach ( $all_plans as $plan ) {
+			if ( intval( self::find_metadata( $plan, 'jetpack_memberships_tier' ) ) === $tier_id ) {
+				$annual_tier = $plan;
+				break;
+			}
+		}
+
+		if ( ! empty( $annual_tier ) ) {
+			$annual_tier_price = floatval( self::find_metadata( $annual_tier, 'jetpack_memberships_price' ) );
+			$valid_plan_ids[]  = $annual_tier->ID;
+		}
+
+		foreach ( $all_plans as $post ) {
+			if ( in_array( $post->ID, $valid_plan_ids, true ) ) {
+				continue;
+			}
+
+			$plan_price    = self::find_metadata( $post, 'jetpack_memberships_price' );
+			$plan_currency = self::find_metadata( $post, 'jetpack_memberships_currency' );
+			$plan_interval = self::find_metadata( $post, 'jetpack_memberships_interval' );
+
+			if ( $plan_price === null || $plan_currency === null || $plan_interval === null ) {
+				// There is an issue with the meta
+				continue;
+			}
+
+			$plan_price = floatval( $plan_price );
+
+			if ( $tier_currency !== $plan_currency ) {
+				// For now, we don't count if there are different currency (not sure how to convert price in a pure JP env)
+				continue;
+			}
+
+			if ( ( $plan_interval === '1 month' && $plan_price >= $tier_price ) ||
+				( $plan_interval === '1 year' && $plan_price >= $annual_tier_price )
```

Comment:
We might also want to make sure that the `$annual_tier_price > 0` here as well, to make sure it got redefined to a real value.